classes/class-wc-rest-connect-shipping-label-controller.php

```diff
@@ -19,28 +19,43 @@ public function post( $request ) {
 		$settings[ 'ship_date' ] = date( 'Y-m-d', time() + 86400 ); // tomorrow
 		$settings[ 'order_id' ] = $order_id;
 
-		$service_names = array();
+		$labels_meta = array();
+		$package_lookup = $this->settings_store->get_package_lookup();
 		foreach ( $settings[ 'packages' ] as $index => $package ) {
-			$service_names[] = $package[ 'service_name' ];
+			$service_name = $package[ 'service_name' ];
 			unset( $package[ 'service_name' ] );
 			$settings[ 'packages' ][ $index ] = $package;
-		}
 
-		$response = $this->api_client->send_shipping_label_request( $settings );
+			$product_names = array();
+			foreach ( $package[ 'products' ] as $product_id ) {
+				$product = wc_get_product( $product_id );
+				if ( $product ) {
+					$product_names[] = $product->get_title();
+				} else {
+					$order = wc_get_order( $order_id );
+					$product_names[] = WC_Connect_Compatibility::instance()->get_product_name_from_order( $product_id, $order );
+				}
+			}
+
+			$box_id = $package[ 'box_id' ];
+			if ( 'individual' === $box_id ) {
+				$package_name = __( 'Individual packaging', 'woocommerce-services' );
+			} else if ( isset( $package_lookup[ $box_id ] ) ) {
+				$package_name = $package_lookup[ $box_id ][ 'name' ];
+			} else {
+				$package_name = __( 'Unknown package', 'woocommerce-services' );
+			}
 
-		if ( is_wp_error( $response ) ) {
-			$error = new WP_Error(
-				$response->get_error_code(),
-				$response->get_error_message(),
-				array( 'message' => $response->get_error_message() )
+			$labels_meta[] = array(
+				'service_name' => $service_name,
+				'product_names' => $product_names,
+				'package_name' => $package_name,
 			);
-			$this->logger->debug( $error, __CLASS__ );
-			return $error;
 		}
 
-		$label_ids = array();
-		$purchased_labels_meta = array();
-		$package_lookup = $this->settings_store->get_package_lookup();
+		$settings[ 'async' ] = true; // TODO: only make it async if the request comes from the Jetpack proxy?
+		$response = $this->api_request( 'send_shipping_label_request', $settings );
```

Comment:
The only real changes needed for an endpoint to start accepting `async requests` are these:
* Replace `$response = $this->api_client->method_name( $arg1, $arg2 );` for `$response = $this->api_request( 'method_name', $arg1, $arg2 );`
* Make sure that no server state is changed **before** calling `$this->api_request()`.
* Make sure that the `async_token` parameter sent by the client will end up in the `$body` of the HTTP request.

Handling program flow using exceptions is a bit hacky, and I don't feel super proud of it, but it allows the individual endpoint handlers to remain virtually unchanged, hiding all the async-related complexity in the parent class.